Inconsistent assignability of enum literal unions vs enum

**Bug Report**

Enum should be assignable to union of its literals (expansion). Mypy usually understands that, but not when union is behind an alias.

**To Reproduce**

[playground](https://mypy-play.net/?mypy=master&python=3.12&flags=strict&gist=4a1a19c716649defc7832b9fa1a03bdf)


```python
from enum import Enum
from typing import Literal

class E(Enum):
    A = 'a'
    B = 'b'
    C = 'c'
    
A = Literal[E.A]
B = Literal[E.B, E.C]

def f(x: A | B) -> None: ...
def f2(x: A | Literal[E.B, E.C]) -> None: ...
def f3(x: Literal[E.A] | B) -> None: ...

def main(x: E) -> None:
    f(x)  # E: Argument 1 to "f" has incompatible type "E"; expected "Literal[E.A] | Literal[E.B, E.C]"  [arg-type]
    f2(x)
    f3(x)  # E: Argument 1 to "f" has incompatible type "E"; expected "Literal[E.A] | Literal[E.B, E.C]"  [arg-type]
```

**Expected Behavior**

Success, 0 errors.

**Actual Behavior**

```
main.py:17: error: Argument 1 to "f" has incompatible type "E"; expected "Literal[E.A] | Literal[E.B, E.C]"  [arg-type]
main.py:19: error: Argument 1 to "f3" has incompatible type "E"; expected "Literal[E.A] | Literal[E.B, E.C]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 1.15.0 and master
- Mypy command-line flags: n/a
- Mypy configuration options from `mypy.ini` (and other config files): n/a
- Python version used: 3.12

I have a patch for this - just a small impl omission. Discovered while looking at #18895.

--- a/mypy/subtypes.py
+++ b/mypy/subtypes.py
@@ -67,6 +67,7 @@ from mypy.types import (
     UnionType,
     UnpackType,
     find_unpack_in_list,
+    flatten_nested_unions,
     get_proper_type,
     is_named_instance,
     split_with_prefix_and_suffix,
@@ -327,7 +328,9 @@ def _is_subtype(
             and isinstance(left, Instance)
             and (left.type.is_enum or left.type.fullname == "builtins.bool")
         ):
-            right = UnionType(mypy.typeops.try_contracting_literals_in_union(right.items))
+            right = UnionType(
+                mypy.typeops.try_contracting_literals_in_union(flatten_nested_unions(right.items))
+            )
             if proper_subtype:
                 is_subtype_of_item = any(
                     is_proper_subtype(orig_left, item, subtype_context=subtype_context)

--- a/mypy/typeops.py
+++ b/mypy/typeops.py
@@ -1069,6 +1069,8 @@ def try_expanding_sum_type_to_union(typ: Type, target_fullname: str) -> ProperTy
 def try_contracting_literals_in_union(types: Sequence[Type]) -> list[ProperType]:
     """Contracts any literal types back into a sum type if possible.
 
+    Requires a flattened union and does not descend into children.
+
     Will replace the first instance of the literal with the sum type and
     remove all others.
 

--- a/test-data/unit/check-literal.test
+++ b/test-data/unit/check-literal.test
@@ -2765,6 +2765,28 @@ reveal_type(x)  # N: Revealed type is "Literal[__main__.Foo.A]"
 reveal_type(y)  # N: Revealed type is "Literal[__main__.Foo.A]"
 [builtins fixtures/tuple.pyi]
 
+[case testLiteralUnionEnumAliasAssignable]
+from enum import Enum
+from typing import Literal, Union
+
+class E(Enum):
+    A = 'a'
+    B = 'b'
+    C = 'c'
+
+A = Literal[E.A]
+B = Literal[E.B, E.C]
+
+def f(x: Union[A, B]) -> None: ...
+def f2(x: Union[A, Literal[E.B, E.C]]) -> None: ...
+def f3(x: Union[Literal[E.A], B]) -> None: ...
+
+def main(x: E) -> None:
+    f(x)
+    f2(x)
+    f3(x)
+[builtins fixtures/tuple.pyi]
+
 [case testStrictEqualityLiteralTrueVsFalse]
 # mypy: strict-equality
 

